### ISSUE TITLE
Add support for additional survey question type "single_select"

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#3760]
When the initial user feedback survey was introduced, scope of the MVP was limited to only support questions of the "rating" type, since that was the only one used for the initial survey. This should now be followed up on, with this issue focusing on the "single_select" question type.
![Screen Shot 2021-07-26 at 3 51 49 PM](https://user-images.githubusercontent.com/3531426/127069257-05074421-4a7a-4c92-902c-4ec2e832403f.png)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A new `SurveyQuestionSingleSelect` JS component should be introduced and integrated into the user feedback survey JS infrastructure, rendering a survey question of type "single_select".
* The UI should look like the above screenshot and follow the "rating" question type where similar (e.g. font, spacing, G icon etc.).
* The available arguments to consider here from the Site Kit Service `/survey/trigger/` API response are fairly similar to those for the "rating" type, as follows:
    * `question.question.answer_choice` contains the list of choices, with `answer_ordinal` (int) and `text` (string), like for the rating question type.
        * In addition, an extra `write_in` boolean property should be supported within each answer choice. If set to `true`, the respective answer should offer a free-text input where the user can type something custom. In that case, the `text` field should be displayed right before the text input. The text input must be non-empty in order to proceed. The maximum character limit should be 100. (This additional free-text input concept is also used in #3761, so this should be coordinated to work similarly, potentially reuse common logic where possible.)
* The data object for answering the question (what needs to be passed to the `answerQuestion` function prop) should look as follows:
    * A single `answer` property, containing an object with the selected `answer_ordinal` (similar to the "rating" question type), and in addition `answer_text` (string with free-text input value, if applicable for the selected answer).
    * Example: `{ answer: { answer_ordinal: 4, answer_text: 'My custom response' } }`
* The new question type should have an example in Storybook where it can be inspected, similar to the existing "rating" type component.

## Implementation Brief
* Create `assets/js/components/surveys/SurveyQuestionSingleSelect.js` which exports the `SurveyQuestionSingleSelect` functional component with the following props (similar to `SurveyQuestionRating`):
	* `question`: Required string
	* `choices`: Required array of objects with the `answer_ordinal`, `text` and `write_in` keys.
	* `answerQuestion`: Required function to be called when a choice has been selected.
	* `dismissSurvey`: Required function to dismiss the survey.
* Create 2 state variables within `SurveyQuestionSingleSelect`:
	* `value`: which is an empty string by default and which will be used for "storing" the selected user option.
	* `writeIn`: which is an empty string by default and which will be used for text input, asssuming that we will ever have 1 write in per question.
* `SurveyQuestionSingleSelect`:
	* should render the question via the `SurveyHeader` component.
	* The choices should be rendered via the `SurveyQuestionSingleSelectChoice` component (more on that below).
	* have a `Next` button powered by the `Button` component sitting at the bottom right of the survey. It should have a callback which submits the selected option for the question.
		* The callback should in turn call the `answerQuestion` prop function, passing an object which has the answer key and having a object as value with `answer_ordinal` as key and the user selected option as value. The user selected ordinal is stored in the `value` state variable. If there is a write in, i.e free text input, via the `writeIn` state variable, it should be added as well to the object, at the same level as `answer_ordinal` and having `answer_text` as value. For e.g `{ answer: { answer_ordinal: 4, answer_text: 'My custom response' } }`
		* If the `value` and `writeIn` variables are empty, the `Next` button should be disabled.
* Create `assets/js/components/surveys/SurveyQuestionSingleSelectChoice.js` which exports the `SurveyQuestionSingleSelectChoice` functional component with the following props:
	* `choice`: Required object having the `answer_ordinal`, `text` and `write_in` keys.
	* `value`: proxied from `SurveyQuestionSelect`
	* `setValue`: proxied from `SurveyQuestionSelect`
	* `writeIn`: proxied from `SurveyQuestionSelect`
	* `setWriteIn`: proxied from `SurveyQuestionSelect`
* `SurveyQuestionSingleSelectChoice`:
	* It should use the `Radio` component to display the choice. `Radio` should have `choice.answer_ordinal` as `value` prop, and `checked` if `value` equals the answer ordinal. The `onClick` and `name` props should be added to the `Radio` component as well.
	* The `onClick` callback should add the selected choice to the `value` variable via the `setValue` function, overwriting any current value each time.
	* If `choice.write_in` is `true`, then render an input text using the `Input` component to the right of the button. The input text should have a max length of `100` characters. The value of the `Input` should be from the `writeIn`.
	* The `Input` should have the `onChange` prop and the callback should update the text value for that answer ordinal via the `setWriteIn` function.
* Create `assets/sass/components/surveys/_googlesitekit-survey-single-select.scss` which will contain all the styles for the `SurveyQuestionSingleSelect` component and its child components.
	* The container wrapping the `Radio` and `Input` in `SurveyQuestionSingleSelectChoice` should have `display: flex` and `flex-wrap` set to `wrap`, so that both components are side by side.
	* The input text should have a `transparent` background as well and same height as the radio button. Font size should be set to `12px`
* Update the `ComponentMap` in `assets/js/components/surveys/CurrentSurvey.js` to map the `SurveyQuestionSingleSelect` component to `single_select`.
* Update `assets/js/components/surveys/__fixtures__/index.js` to include sample payload for single select survey types.
* Add stories for `SurveyQuestionSingleSelect` in `assets/js/components/surveys/SurveyQuestionSingleSelect.stories.js`.

### Test Coverage
* Tests to be added for the `SurveyQuestionSingleSelect` component.
* Add new tests within `assets/js/components/surveys/CurrentSurvey.test.js` for the new question type.

### Visual Regression Changes
* N/A

## QA Brief

Check that https://google.github.io/site-kit-wp/storybook/develop/path=/story/components-surveys--survey-question-single-select-story works and behaves as per AC

## Changelog entry

* Add a new "single select" option to User Input Surveys.

**Comment by asvinb** [comment]
@felixarntz @aaemnnosttv I have a few questions based on the screenshot:

- In the screenshot, we have the `Next` button which I presume submits a user selected answer. However in our current setup, we don't have the `Next` button. Rather the answer is submitted as soon as (after 350ms to be exact) the user selects an answer. Can you confirm we will keep our existing functionality?
- If `choice.write_in` is `true`, when is the input text visible? Is it upon component mount or it's visible only if the user selects the related choice?
- Then, how is the answer submitted once the user has finished typing? Is it upon hitting the `Enter` key?

Based on the IB I added, this is how the survey will potentially look:
![image](https://user-images.githubusercontent.com/31818/127335640-0f284ad1-9e6f-41ed-ac3e-d97cc1008c02.png)

I'am wondering whether the input text should be below the option rather than next to it 🤔

**Comment by felixarntz** [comment]
@asvinb Great questions.

* The "Next" button should be there and required - it is only the "rating" question type specifically that does not require it, which is according to how Google's own survey system handles it.
* It should always be visible, so that it's clear the user can enter something when making this choice.
* See above, everything is only submitted once clicking "Next".

**Comment by asvinb** [comment]
@felixarntz @aaemnnosttv I've updated the IB based on your comments. However I think that maybe we should create a single component, for e.g `SurveyQuestionSelect` which caters for both, single and multi select. The IB caters for that. This is the same approach we are using for user input. Let me know what you think, because instead of duplicating most of the functionality, with some small tweaks, we can achieve both single and multi selects within the same component.
Also, I think we can keep it simple for single selects to use the `Radio` component. This will differ slightly from what we have in the JPG but i think it's better from a user perspective and works well with our current component. This is potentially how it'll look:

![image](https://user-images.githubusercontent.com/31818/127484695-3c78e189-3294-4f9d-a877-43b3ce84cf2b.png)
![image](https://user-images.githubusercontent.com/31818/127484717-f2af6b8d-d296-4195-a1ab-1c48438f3fcf.png)

Let me know what you think.

**Comment by aaemnnosttv** [comment]
Thanks @asvinb – I really like the idea of using radio and checkbox inputs to make the selections (and number of selections) much more clear as well as accessible, although this is a deviation from the design so we'd need to get approval, but +1 from me. cc: @felixarntz @marrrmarrr 

Regarding the use of a single component, I think that is of course technically possible but in this case the survey infrastructure is designed to [map each question's type to a single component](https://github.com/google/site-kit-wp/blob/47b723d1416fa48b4e9df73b2970e868e1ccece8/assets/js/components/surveys/CurrentSurvey.js#L201-L215). So perhaps there could be a shared/variable component inside, but we still want/need separate components in the end for this mapping as the props going in are fixed since this is entirely API-driven. I would prefer to keep them separate though unless there are particularly substantial gains from building as one – if there are, please clarify 😄

**Comment by felixarntz** [comment]
@asvinb @aaemnnosttv +1 to also showing the radio buttons next to the selections for "single_select" - we can proceed with that.

Regarding a single component, I agree with @aaemnnosttv - how about the following:
* We introduce `SurveyQuestionSelect` as a base component, with a few extra props (see below).
* We introduce `SurveyQuestionSingleSelect` (and through #3761 later) `SurveyQuestionMultiSelect`, both of which mostly use `SurveyQuestionSelect` under the hood.
* It may be enough to have a boolean `multiple` prop as the only extra piece and handle the few differences in there (checkboxes instead of radio buttons, sending multiple answer selections instead of a single answer selection, etc.).

If we go with that, it may be more appropriate to even combine #3760 and #3761? If we keep them separate, I'd be worried that the abstraction for `SurveyQuestionSelect` eventually may not satisfy both use-cases. If we keep them together, this would be ensured.

**Comment by asvinb** [comment]
Thanks @felixarntz @aaemnnosttv  👍 for the radio buttons.

I think introducing the `SurveyQuestionSelect` sounds good and I've updated the IB to cater for that. Also, tackling #3760 and #3761 together makes sense. The IB reflects that.

I've kept the old IB as well just in case we want to keep the components separate. Once the IB is approved, I'll remove the old one.

**Comment by danielgent** [comment]
What's the benefit with creating a single Frankenstein component though?

More logic? More room for bugs? Less clear component API?

If we have common code between the two components then we should extract them as util functions or other child components.

@asvinb Please provide me with a reason as to why they should be combined?  :smile: 

I've already built the multiselect as well so the work is already done.  But the state is different, the inputs are different, the data shape it calls the callback with is different :shrug:

**Comment by aaemnnosttv** [comment]
@felixarntz after discussing this a bit more with @asvinb, we've opted to go with separate components without a shared base component. The reason being mostly due to the cost of added complexity outweighing the benefit of making the two into a single component. The survey components are already a bit complex as it is with the structure of questions and answers so I think it's best to keep each component on the simpler side rather than creating a hybrid that can do both. This also gives us more flexibility with both should we want to make changes later since the components will be independent.

Let me know if you'd still prefer to go with the abstraction, otherwise we can continue with this one as the single component as initially scoped 👍

**Comment by felixarntz** [comment]
@aaemnnosttv That's fair - okay from my end.

**Comment by aaemnnosttv** [comment]
@asvinb please update the IB to only cover the single component as originally scoped. Thanks!

**Comment by asvinb** [comment]
Thanks @felixarntz @aaemnnosttv IB updated

**Comment by aaemnnosttv** [comment]
Thanks @asvinb – the IB looks good and is quite thorough – just a few things to iterate on:

> * Create 2 state variables within `SurveyQuestionSingleSelect`:
>	* `values`: which is an empty array by default. Selected options will be pushed to this variable
>	* `writeIns`: which is an empty object by default. Text inputs will be added to this object, with the `answer_ordinal` as key (in case we have more than 2 write ins for a question).

This reads like it's accounting for multiple possible answers but the question type only supports one 🙂  It's technically possible that each question could support a write-in value, but only one choice +write-in can be used as the answer. You show this again in the example answer object (`{ answer: [ { answer_ordinal: 2 }, { answer_ordinal: 4, answer_text: 'My custom response' } ] }`) which doesn't match the structure of the example in the ACs.

> 	* `values`: proxied from `SurveyQuestionSelect`
>	* `setValues`: proxied from `SurveyQuestionSelect`
>	* `writeIns`: proxied from `SurveyQuestionSelect`
>	* `setWriteIns`: proxied from `SurveyQuestionSelect`

Is this not referring to the abstraction from the old IB?

> * No new tests to be added.

Given the complexity around these components this is something we'll definitely want to test well.

@tofumatt did the original work on the main/current survey components and has an initial suite started for `CurrentSurvey` which tests everything working together which should be extended with additional tests for each new question type (wrapped in its own `describe`). This is particularly important for the new question types and the functionality of the "Next" button which decouples the submission of the answer and the selection of the choice as this did not apply to the initial rating type. These tests should be of a similar level of abstraction to what is there now, but grouped by the question type.

Additional lower-level unit tests for the component are optional but could be added separately (i.e. `SurveyQuestionSingleSelect.test.js`) if deemed useful.

**Comment by asvinb** [comment]
Thanks @aaemnnosttv 🤦 Yeah there are still some old references to having multiple values. I updated the IB. Can you take a look once again?

Thanks!

**Comment by tofumatt** [comment]
IB ✅  from me.

I'll leave @aaemnnosttv assigned to take a look as well, but I think this looks good. 👍🏻

**Comment by danielgent** [comment]
I think the implementation brief is missing complexity around the handling of `write_in` 

If the user selects an option that has `write_in`, what happens if they enter text and then choose another value? Should this value be reset? (I'm not going to do this as there is no mention of this)

_"If there is a write in, i.e free text input, via the writeIn state variable,  it should be added as well to the object"_
 Should be more specific: _"If the currently selected option has a `write_in` then..."_ as we would expect `write_in` to be empty mostly (as most options don't have `write_in`)


_"If the value and writeIn variables are empty, the Next button should be disabled."_
Should be "If the current value has a text input, and that input is empty".  Because for most values it's expected that `write_in` is empty


Also the IB assumes only one option will have `write_in` as true, otherwise we're gonna see some very unexpected behavior here!

**Comment by wpdarren** [comment]
@danielgent @tofumatt I have found [the storybook](https://google.github.io/site-kit-wp/storybook/develop/?path=/story/components-surveys--survey-question-single-select-story) (the link in the QAB 404's) and can see how the single select form works but would prefer to see this on a test site, so I can try break it 😃   E.g. on Storybook when I click on the dismiss icon nothing happens, so testing on here is not reflecting the behaviour I'd expect.  

Would it be possible to do something in the tester plugin, or offer instructions on how I can test this in live environment.

**Comment by danielgent** [comment]
@wpdarren - We should probably include these new controls inside the `CurrentSurvey` stories e.g. the Multiple Questions one https://google.github.io/site-kit-wp/storybook/develop/?path=/story/components-surveys-currentsurvey--survey-multiple-questions-story

Good point that these cannot be QA'd just from the story showing the new component individually

**Comment by wpdarren** [comment]
### QA Update:  Pass ✅ 

- When activating User Feedback on the Tester plugin, and using the code supplied,  a new `SurveyQuestionSingleSelect JS component` is implemented, rendering a survey question of type `single_select.` with radio buttons.  - [Screenshot](https://user-images.githubusercontent.com/73545194/129588280-f5338d7c-cf3c-4ab1-b46d-2e26ec766e63.png)
- The UI should looks like the above screenshot and follows the `rating` question type formating, i.e shadow, etc. 
- Within the survey question form there is a single question, with a list of choices. 
- Within the survey question form there is an `other` option with text field. No more than 100 characters can be added.
- For the `other` option the submit button is not active until text is added.  -  [Screenshot](https://user-images.githubusercontent.com/73545194/129587015-117c3c74-e006-426e-b8e7-299e6a2cfc41.png)
- Tested the survey question form on supported browsers and smaller screen sizes.
